router/handle.go

```diff
@@ -242,6 +249,14 @@ func (rt *Handle) pickup(ctx context.Context, partition string, workers []*worke
 
 	flush()
 	limitsReached = iteratorStats.LimitsReached
+	discardedRatio := float64(iteratorStats.DiscardedJobs) / float64(iteratorStats.TotalJobs)
+
+	// If the discarded ratio is greater than the penalty threshold,
+	// sleep for a while to avoid having a loop running continuously without producing events
+	if limitsReached && discardedRatio > rt.reloadableConfig.failingJobsPenaltyThreshold {
+		limiterEnd()
+		_ = misc.SleepCtx(ctx, rt.reloadableConfig.failingJobsPenaltySleep)
+	}
 	rt.pipelineDelayStats(partition, firstJob, lastJob)
```

Comment:
Would it make to move the logic in the function that calls pickup, or in a wrapper function, so that `limiterEnd()` remains on defer?

Reply:
since you noticed I will do it properly in rudder-go-kit :)

Reply:
https://github.com/rudderlabs/rudder-go-kit/pull/71